File: src/core/pluginRoot.ts
/**
 * Plugin root resolver — 4-strategy chain that locates the
 * harness-boot plugin source at runtime (F-087 port of
 * `scripts/core/plugin_root.py`).
 *
 * Strategy order — first hit wins:
 *
 *   A. `$PATH` scan for a `/plugins/<dir>/bin` segment whose parent
 *      directory contains a `plugin.json` with a matching `name`.
 *   B. `~/.claude/plugins/installed_plugins.json` `installPath` for
 *      the requested plugin (the path must exist on disk).
 *   C. `~/.claude/settings.json` `extraKnownMarketplaces.<mp>.source.path`
 *      combined with the matching `marketplace.json` plugin source.
 *   D. (Failure) — throw {@link PluginRootError} with the attempt
 *      trace.
 *
 * The implementation is a direct port of the Python module — every
 * decision (PATH parsing, name matching, fallback ordering, error
 * message string) is preserved verbatim so cross-runtime tooling can
 * compare logs without translation tables.
 *
 * @module pluginRoot
 */

import {existsSync, readFileSync, realpathSync, statSync} from 'node:fs';
import {homedir} from 'node:os';
import {delimiter as pathDelimiter, join, resolve as resolvePath} from 'node:path';

/**
 * Thrown by {@link resolve} when every strategy has missed.
 *
 * The `message` matches the Python implementation byte-for-byte
 * (Korean phrasing preserved) so log scrapers can match either
 * runtime's output with a single regex.
 */
export class PluginRootError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'PluginRootError';
  }
}

/**
 * Successful resolution result plus the trace of which strategies
 * were tried in what order.
 *
 * `root` is the absolute, symlink-resolved plugin directory.
 * `strategy` is one of `'A:path-bin'`, `'B:registry'`,
 * `'C:marketplace-source'`. `attempts` is the human-readable trace —
 * each entry is one of `'A path-bin: hit'`, `'A path-bin: miss'`, …
 */
export interface Resolution {
  root: string;
  strategy: 'A:path-bin' | 'B:registry' | 'C:marketplace-source';
  attempts: string[];
}

/** Reads + parses `plugin.json` for a candidate root, or returns null. */
function loadPluginJson(root: string): Record<string, unknown> | null {
  const manifest = join(root, '.claude-plugin', 'plugin.json');
  try {
    if (!statSync(manifest).isFile()) {
      return null;
    }
    const raw = readFileSync(manifest, 'utf-8');
    const parsed: unknown = JSON.parse(raw);
    if (parsed === null || typeof parsed !== 'object' || Array.isArray(parsed)) {
      return null;
    }
    return parsed as Record<string, unknown>;
  } catch {
    return null;
  }
}

/** Mirrors `Path.resolve()` from Python — symlinks, `..` collapsed. */
function tryRealpath(p: string): string | null {
  try {
    return realpathSync(p);
  } catch {
    return null;
  }
}

/** Returns true iff `p` exists and is a directory. */
function isDirectory(p: string): boolean {
  try {
    return statSync(p).isDirectory();
  } catch {
    return false;
  }
}

/**
 * Strategy A — `$PATH` scan.
 *
 * Looks for entries that contain `/plugins/` and end in `/bin`. The
 * parent directory of such an entry is the candidate plugin root; we
 * accept it only when its `plugin.json.name` matches the requested
 * plugin name.
 *
 * Exported under the underscore-prefixed alias for the parity test —
 * the Python module exposes the same internals.
 */
export function _strategyPathBin(pluginName: string): string | null {
  const pathEnv = process.env.PATH ?? '';
  for (const entry of pathEnv.split(pathDelimiter)) {
    if (!entry) {
      continue;
    }
    if (!entry.includes('/plugins/')) {
      continue;
    }
    const trimmed = entry.replace(/\/+$/, '');
    if (!trimmed.endsWith('/bin')) {
      continue;
    }
    // The plugin root is the parent of the bin directory.
    const candidate = resolvePath(entry, '..');
    const real = tryRealpath(candidate);
    if (real === null) {
      continue;
    }
    const manifest = loadPluginJson(real);
    if (manifest && manifest['name'] === pluginName) {
      return real;
    }
  }
  return null;
}

/**
 * Strategy B — `installed_plugins.json` lookup.
 *
 * Inspects the `~/.claude/plugins/installed_plugins.json` registry.
 * For each `<plugin>@<marketplace>` key prefixed with the requested
 * plugin name we read the first entry's `installPath` and return it
 * if it exists on disk.
 *
 * NEW-44 — the registry can record an `installPath` that no longer
 * exists for directory-type marketplaces; the existence check guards
 * against returning a phantom path.
 */
export function _strategyRegistry(pluginName: string): string | null {
  const registry = join(homedir(), '.claude', 'plugins', 'installed_plugins.json');
  let parsed: unknown;
  try {
    if (!statSync(registry).isFile()) {
      return null;
    }
    parsed = JSON.parse(readFileSync(registry, 'utf-8'));
  } catch {
    return null;
  }
  if (parsed === null || typeof parsed !== 'object' || Array.isArray(parsed)) {
    return null;
  }
  const plugins = (parsed as Record<string, unknown>)['plugins'];
  if (plugins === null || typeof plugins !== 'object' || Array.isArray(plugins)) {
    return null;
  }
  const prefix = `${pluginName}@`;
  for (const [key, entries] of Object.entries(plugins as Record<string, unknown>)) {
    if (!key.startsWith(prefix)) {
      continue;
    }
    if (!Array.isArray(entries) || entries.length === 0) {
      continue;
    }
    const head = entries[0];
    if (head === null || typeof head !== 'object' || Array.isArray(head)) {
      continue;
    }
    const installPath = (head as Record<string, unknown>)['installPath'];
    if (typeof installPath !== 'string' || installPath.length === 0) {
      continue;
    }
    const expanded = expandHome(installPath);
    if (isDirectory(expanded)) {
      return tryRealpath(expanded) ?? expanded;
    }
  }
  return null;
}

/**
 * Strategy C — directory-type marketplace fallback.
 *
 * When the registry's `installPath` does not exist (NEW-44) we walk
 * the user's `extraKnownMarketplaces`, pick the directory-type ones,
 * read each marketplace's `marketplace.json`, and combine the
 * marketplace path with the plugin's relative `source` field.
 */
export function _strategyMarketplaceSource(pluginName: string): string | null {
  const settings = join(homedir(), '.claude', 'settings.json');
  let parsed: unknown;
  try {
    if (!statSync(settings).isFile()) {
      return null;
    }
    parsed = JSON.parse(readFileSync(settings, 'utf-8'));
  } catch {
    return null;
  }
  if (parsed === null || typeof parsed !== 'object' || Array.isArray(parsed)) {
    return null;
  }
  const marketplaces = (parsed as Record<string, unknown>)['extraKnownMarketplaces'];
  if (
    marketplaces === null ||
    typeof marketplaces !== 'object' ||
    Array.isArray(marketplaces)
  ) {
    return null;
  }
  for (const mpDef of Object.values(marketplaces as Record<string, unknown>)) {
    if (mpDef === null || typeof mpDef !== 'object' || Array.isArray(mpDef)) {
      continue;
    }
    const src = (mpDef as Record<string, unknown>)['source'];
    if (src === null || typeof src !== 'object' || Array.isArray(src)) {
      continue;
    }
    if ((src as Record<string, unknown>)['source'] !== 'directory') {
      continue;
    }
    const pathStr = (src as Record<string, unknown>)['path'];
    if (typeof pathStr !== 'string' || pathStr.length === 0) {
      continue;
    }
    const mpRoot = expandHome(pathStr);
    if (!isDirectory(mpRoot)) {
      continue;
    }
    const mpManifestPath = join(mpRoot, '.claude-plugin', 'marketplace.json');
    let mpData: unknown;
    try {
      if (!statSync(mpManifestPath).isFile()) {
        continue;
      }
      mpData = JSON.parse(readFileSync(mpManifestPath, 'utf-8'));
    } catch {
      continue;
    }
    if (mpData === null || typeof mpData !== 'object' || Array.isArray(mpData)) {
      continue;
    }
    const pluginEntries = (mpData as Record<string, unknown>)['plugins'];
    if (!Array.isArray(pluginEntries)) {
      continue;
    }
    for (const entry of pluginEntries) {
      if (entry === null || typeof entry !== 'object' || Array.isArray(entry)) {
        continue;
      }
      if ((entry as Record<string, unknown>)['name'] !== pluginName) {
        continue;
      }
      const pluginSrc = (entry as Record<string, unknown>)['source'] ?? './';
      if (typeof pluginSrc !== 'string') {
        // Remote sources (objects) — out of scope for directory-type fallback.
        continue;
      }
      const candidate = resolvePath(mpRoot, pluginSrc);
      if (isDirectory(candidate)) {
        return tryRealpath(candidate) ?? candidate;
      }
    }
  }
  return null;
}

/**
 * Expands a leading `~` or `~/` to the current user's home directory.
 *
 * Mirrors Python's `os.path.expanduser` for the common cases — only
 * leading `~` is expanded, embedded `~` characters remain literal.
 */
function expandHome(p: string): string {
  if (p === '~') {
    return homedir();
  }
  if (p.startsWith('~/')) {
    return join(homedir(), p.slice(2));
  }
  return p;
}

/**
 * Runs the 4-strategy chain and returns the first hit.
 *
 * @param pluginName - Plugin manifest name to match (default
 *   `'harness'`, matching the Python source).
 * @returns Successful {@link Resolution}.
 * @throws {@link PluginRootError} when every strategy has missed.
 */
export function resolve(pluginName: string = 'harness'): Resolution {
  const attempts: string[] = [];

  let root = _strategyPathBin(pluginName);
  attempts.push(`A path-bin: ${root ? 'hit' : 'miss'}`);
  if (root) {
    return {root, strategy: 'A:path-bin', attempts};
  }

  root = _strategyRegistry(pluginName);
  attempts.push(`B registry: ${root ? 'hit' : 'miss'}`);
  if (root) {
    return {root, strategy: 'B:registry', attempts};
  }

  root = _strategyMarketplaceSource(pluginName);
  attempts.push(`C mp-source: ${root ? 'hit' : 'miss'}`);
  if (root) {
    return {root, strategy: 'C:marketplace-source', attempts};
  }

  throw new PluginRootError(
    `플러그인 '${pluginName}' 루트 해석 실패 — attempts: ${attempts.join(', ')}`,
  );
}

/** Predicate convenience — returns true when {@link resolve} would succeed. */
export function isResolvable(pluginName: string = 'harness'): boolean {
  try {
    resolve(pluginName);
    return true;
  } catch (err) {
    if (err instanceof PluginRootError) {
      return false;
    }
    throw err;
  }
}

/**
 * Re-exported for tests that need to verify the existsSync semantics
 * Python's `Path.is_dir()` uses without re-importing node:fs.
 */
export const _internals = {
  loadPluginJson,
  expandHome,
  isDirectory,
  existsSync,
};

File: legacy/scripts/core/plugin_root.py
#!/usr/bin/env python3
"""
plugin_root.py — harness-boot 플러그인 루트 경로 해석 (NEW-37 + NEW-44)

`commands/init.md §2` 의 4-전략 체인을 Python 함수로 재사용 가능하게 추출.
sync.py · work.py 등 다른 스크립트에서 플러그인 내부 자산을 찾을 때 공통 사용.

사용:
  python3 scripts/plugin_root.py                  # 경로 1 줄 출력
  python3 scripts/plugin_root.py --json           # 전략 추적 JSON
  python3 scripts/plugin_root.py --plugin harness # 기본값 "harness"

전략 (첫 성공이 이김):
  A. $PATH 에서 `/plugins/.*/bin$` 패턴 찾아 역산 + plugin.json.name 매칭
  B. ~/.claude/plugins/installed_plugins.json 의 installPath (실존 시)
  C. ~/.claude/settings.json 의 extraKnownMarketplaces.<mp>.source.path + marketplace.json 의 plugin source
  D. 실패 시 PluginRootError

외부 의존: pyyaml (선택 — marketplace.json 이 yaml 일 경우).
"""

from __future__ import annotations

import argparse
import json
import os
import sys
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any


class PluginRootError(Exception):
    """플러그인 루트 해석 실패."""


@dataclass
class Resolution:
    """성공한 해석 결과 + 어떤 전략이 맞았는지 추적."""

    root: Path
    strategy: str
    attempts: list[str] = field(default_factory=list)


def _load_plugin_json(root: Path) -> dict | None:
    manifest = root / ".claude-plugin" / "plugin.json"
    if not manifest.is_file():
        return None
    try:
        with manifest.open("r", encoding="utf-8") as f:
            return json.load(f)
    except (OSError, json.JSONDecodeError):
        return None


def _strategy_path_bin(plugin_name: str) -> Path | None:
    """전략 A — $PATH 에서 /plugins/<dir>/bin 조각을 찾아 부모 디렉터리를 반환.

    plugin.json.name 이 plugin_name 과 매칭되는 것만 채택.
    """
    path_env = os.environ.get("PATH", "")
    for entry in path_env.split(os.pathsep):
        if not entry:
            continue
        # /plugins/ 를 포함하고 /bin 으로 끝나는 항목
        if "/plugins/" not in entry or not entry.rstrip("/").endswith("/bin"):
            continue
        candidate = Path(entry).parent
        try:
            resolved = candidate.resolve()
        except (OSError, RuntimeError):
            continue
        manifest = _load_plugin_json(resolved)
        if manifest and manifest.get("name") == plugin_name:
            return resolved
    return None


def _strategy_registry(plugin_name: str) -> Path | None:
    """전략 B — installed_plugins.json 의 installPath 가 실존할 때."""
    registry = Path.home() / ".claude" / "plugins" / "installed_plugins.json"
    if not registry.is_file():
        return None
    try:
        with registry.open("r", encoding="utf-8") as f:
            data = json.load(f)
    except (OSError, json.JSONDecodeError):
        return None
    plugins = data.get("plugins", {}) if isinstance(data, dict) else {}
    for key, entries in plugins.items():
        if not key.startswith(f"{plugin_name}@"):
            continue
        if not isinstance(entries, list) or not entries:
            continue
        install_path = entries[0].get("installPath")
        if not install_path:
            continue
        path = Path(os.path.expanduser(install_path))
        if path.is_dir():
            return path.resolve()
    return None


def _strategy_marketplace_source(plugin_name: str) -> Path | None:
    """전략 C — directory-type marketplace 의 source.path 에서 plugin source 결합.

    NEW-44: installPath 가 실존하지 않을 때의 fallback.
    """
    settings = Path.home() / ".claude" / "settings.json"
    if not settings.is_file():
        return None
    try:
        with settings.open("r", encoding="utf-8") as f:
            data = json.load(f)
    except (OSError, json.JSONDecodeError):
        return None

    marketplaces = data.get("extraKnownMarketplaces", {})
    if not isinstance(marketplaces, dict):
        return None

    for mp_name, mp_def in marketplaces.items():
        if not isinstance(mp_def, dict):
            continue
        src = mp_def.get("source", {})
        if src.get("source") != "directory":
            continue
        path_str = src.get("path")
        if not path_str:
            continue
        mp_root = Path(os.path.expanduser(path_str))
        if not mp_root.is_dir():
            continue
        mp_manifest = mp_root / ".claude-plugin" / "marketplace.json"
        if not mp_manifest.is_file():
            continue
        try:
            with mp_manifest.open("r", encoding="utf-8") as f:
                mp_data = json.load(f)
        except (OSError, json.JSONDecodeError):
            continue
        for plugin_entry in mp_data.get("plugins", []):
            if plugin_entry.get("name") != plugin_name:
                continue
            plugin_src = plugin_entry.get("source", "./")
            if not isinstance(plugin_src, str):
                continue  # 원격 source 는 skip
            candidate = (mp_root / plugin_src).resolve()
            if candidate.is_dir():
                return candidate
    return None


def resolve(plugin_name: str = "harness") -> Resolution:
    """4-전략 체인으로 플러그인 루트 해석. 실패 시 PluginRootError."""
    attempts: list[str] = []

    root = _strategy_path_bin(plugin_name)
    attempts.append(f"A path-bin: {'hit' if root else 'miss'}")
    if root:
        return Resolution(root=root, strategy="A:path-bin", attempts=attempts)

    root = _strategy_registry(plugin_name)
    attempts.append(f"B registry: {'hit' if root else 'miss'}")
    if root:
        return Resolution(root=root, strategy="B:registry", attempts=attempts)

    root = _strategy_marketplace_source(plugin_name)
    attempts.append(f"C mp-source: {'hit' if root else 'miss'}")
    if root:
        return Resolution(
            root=root, strategy="C:marketplace-source", attempts=attempts
        )

    raise PluginRootError(
        f"플러그인 '{plugin_name}' 루트 해석 실패 — "
        f"attempts: {', '.join(attempts)}"
    )


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(
        description="Resolve the harness-boot plugin root using the 4-strategy chain"
    )
    parser.add_argument("--plugin", default="harness", help="plugin name (default: harness)")
    parser.add_argument(
        "--json", action="store_true", help="emit JSON with strategy trace"
    )
    args = parser.parse_args(argv)

    try:
        result = resolve(args.plugin)
    except PluginRootError as e:
        print(f"error: {e}", file=sys.stderr)
        return 4

    if args.json:
        json.dump(
            {
                "root": str(result.root),
                "strategy": result.strategy,
                "attempts": result.attempts,
            },
            sys.stdout,
            indent=2,
            ensure_ascii=False,
        )
        print()
    else:
        print(result.root)
    return 0


if __name__ == "__main__":
    sys.exit(main())
